src/core/qgsfields.cpp

```diff
@@ -422,3 +422,10 @@ QDataStream &operator>>( QDataStream &in, QgsFields &fields )
   }
   return in;
 }
+
+bool QgsFields::isFieldReadOnly( int fieldIdx ) const
+{
```

Comment:
Aren't joined fields editable in some circumstances?

Reply:
I'm not familiar with that functionality in QGIS. I've just ported the existing test that was repeated at a few places.

Reply:
Ok, it's messy. There's special handling in the attribute table/form which overrides the default read only behaviour. See eg https://github.com/qgis/QGIS/blob/master/src/gui/attributetable/qgsattributetabledelegate.cpp#L90.

But to add to the confusion, there's also QgsEditFormConfig::readOnly. 

So I think some documentation clarifications are needed here. Eg:

- QgsField::isReadOnly / setReadOnly : "This indicates that the field can never be edited, due to restrictions at the data provider level. Fields which return FALSE to isReadOnly() may still be considered as non-editable if higher level components (such as the configuration in QgsEditFormConfig) have manually marked the field as non-editable".
- QgsEditFormConfig::readOnly / setReadOnly: "\warning A field which has been marked as editable in QgsEditFormConfig may still be treated as read only if the corresponding field returns TRUE to QgsField::isReadOnly, which indicates that editing is blocked by a data provider level restriction."
- QgsFields::isFieldReadOnly: "Fields which return FALSE to isFieldReadOnly() may still be considered as non-editable in the QGIS application if higher level components (such as the configuration in QgsEditFormConfig) have manually marked the field as non-editable".
- QgsFields::isFieldReadOnly: "\warning A field with OriginJoin may be editable via the QgsVectorLayer edit methods if the corresponding QgsVectorLayerJoinInfo is editable and the corresponding field from the joined layer is not marked as read only." (Also probably should be added to QgsField::isReadOnly too)

There's also a confusing mix of "isReadOnly" vs "readOnly" for the getters. We should ultimately decide on one and add aliases to keep consistency across the api. I'd lean toward isReadOnly...